src/Math-Matrix/PMTensor.class.st

```diff
@@ -0,0 +1,152 @@
+Class {
+	#name : #PMTensor,
+	#superclass : #Object,
+	#instVars : [
+		'array',
+		'shape',
+		'first',
+		'strides'
+	],
+	#category : #'Math-Matrix'
+}
+
+{ #category : #'instance creation' }
+PMTensor class >> fromNestedArray: anArray [
+
+	^ self new array: anArray flattened withShape: (self shape:anArray)
+]
+
+{ #category : #'instance creation' }
+PMTensor class >> newWith: anInteger [
+
+	^ self new array: {anInteger} withShape: #(  )
+]
+
+{ #category : #accessing }
+PMTensor class >> shape: anArray [
+
+anArray isArray ifFalse:[^#()]
+ifTrue:[
+	^ {anArray size}, (self shape: anArray first)
+	]
+
+]
+
+{ #category : #comparing }
+PMTensor >> = aTensor [
+
+	^ array = aTensor array & (first = aTensor first)
+	  & (strides = aTensor strides) & (shape = aTensor shape)
+]
+
+{ #category : #accessing }
+PMTensor >> array [ 
+^array
+]
+
+{ #category : #'as yet unclassified' }
+PMTensor >> array: aFlatArray withShape: aShape [
+
+	array := aFlatArray.
+	shape := aShape copy.
+	self updateFirst.
+	shape ifNotEmpty: [ self updateStrides]
+]
+
+{ #category : #accessing }
+PMTensor >> first [ 
+^first
+]
+
+{ #category : #'primitives - file' }
+PMTensor >> flattenedIndexOf: coords [
+
+	| position |
+	position := 1.
+	coords withIndexDo: [ :elt :i | 
+		position := elt - 1 * (strides at: i) + position ].
+	^ position
+]
+
+{ #category : #'as yet unclassified' }
+PMTensor >> fromNestedArray: aFlatArray withShape: aShape [
+
+	array := aFlatArray.
+	shape := aShape copy.
+	self updateFirst.
+	shape ifNotEmpty: [ self updateStrides]
+]
+
+{ #category : #public }
+PMTensor >> get: coords [
+
+	| position |
+	position := self flattenedIndexOf: coords.
+	^ array at: position
+]
+
+{ #category : #accessing }
+PMTensor >> rank [
+	^ shape size
+]
+
+{ #category : #'as yet unclassified' }
+PMTensor >> reshape: aNewShape [
```

Comment:
Normally you can contract tensors (check the literature) such that rank 2 tensors become rank 0 eg trace. Is this the intent here?

Reply:
no this this is reshape like in Numpy: https://numpy.org/doc/stable/reference/generated/numpy.reshape.html
Changing the dimension of the n-dimensional arrays without changing the contents.

Reply:
Ah, so we're copying numpy completely. Non blocking, but not sure if that's wise choice. Is the code quality is good? If it is, fine, go ahead. If not, then do we want those practices brought over?

Reply:
We are not copying completely. This is more an inspiration especially for the internal implementation as one 1D array

Reply:
At this point I would advise we choose wisely.

Reply:
Bias warning: I was part of the team that designed the basic NumPy Python API (back when it was called `Numeric`).

The basics of NumPy were heavily inspired by APL, and even 25 years later I still believe that was a good choice, and so I think we should follow the same path here. Over the years, the NumPy API got bloated, so I definitely don't recommend to follow NumPy everywhere.

Your reference to contraction points to the real problem: it's not a good idea to use the term "tensor" for what is really an N-dimensional array without any implied semantics.

Reply:
I am inclined to agree. Tensors are have a defined algebra and if this isn't a Tensor, then let's correct that. I agree that we should use Numpy terminology everywhere (where the naming is clear and domain-driven, we should).

Generally, where we can let's choose the names from Mathematics where that is possible.

Reply:
@khinsen What is APL?

Reply:
The very first interactive language based on N-dimensional arrays. Developed by Ken Iverson at IBM, first published in the 1960s. Details: https://en.wikipedia.org/wiki/APL_(programming_language)

Reply:
Thanks. @hemalvarambhia what would be the appropriate name for this class ?

Reply:
How about MultidimensionalArray ? I get the feeling this may be what @khinsen meant

Reply:
Thank you. I think I need to discuss it with @khinsen and @SergeStinckwich to make the choice.

Reply:
``PMArray`` is not informative enough I guess, ``PMNDArray`` (n-dimensional array) is better.